### PR TITLE
Fix embind to work (more) on MINIMAL_RUNTIME.

### PR DESCRIPTION
**Pull request by juj**
Fix embind to work (more) on MINIMAL_RUNTIME.

Remove old super-ancient pre-asm.js binding code from embind.

Helps towards #10386, though not sure if it will resolve all the issues seen there.

### CODE REVIEW

File: src/deps_info.json [hunk 1]
```diff
@@ -94,5 +94,8 @@
   "_embind_register_std_string": ["malloc", "free"],
   "_embind_register_std_wstring": ["malloc", "free"],
   "__syscall192": ["emscripten_builtin_memalign"],
-  "pthread_create": ["malloc", "free", "emscripten_main_thread_process_queued_calls"]
+  "pthread_create": ["malloc", "free", "emscripten_main_thread_process_queued_calls"],
+  "$getTypeName": ["free"],
+  "_embind_register_std_string": ["free"],
+  "_embind_register_std_wstring": ["free"]
```
Comment (juj) [inline src/deps_info.json]: @kripken: For some reason I still can't get my head around the deps_info.json behavior. :/ 

E.g. `other.test_embind` calls `_embind_register_std_string` directly from C/C++ code (`__embind_register_native_and_builtin_types` in `bind.cpp`), and `_embind_register_std_string` depends on `free`, so I put in a dependency `"_embind_register_std_string": ["free"],` here, but even then for some reason the function free is not getting included to the build, hence I need an explicit `shared.Settings.EXPORTED_FUNCTIONS += ['_free']` above.

Comment (kripken) [reply src/deps_info.json]: That seems like it should work. I wonder if there is something special about embind linking perhaps?

Should be easy to debug, `add_back_deps` is where the simple logic for deps_info is done, in tools/system_libs.py. Verifying that symbol is undefined, and in the list of symbols scanned there, should show what is going wrong.

Comment (juj) [reply src/deps_info.json]: Still not sure I understand what's going on.. The test program does use `_embind_register_std_string` for example, but it is not present in `undefs`. Adding the dependency to `free` to a few more functions that do also flow a call to `free` resolves the issue. Perhaps that is enough of a coverage to make sure it gets picked up always.

Comment (kripken) [reply src/deps_info.json]: Was this resolved? I don't see `_free` added to EXPORTED_FUNCTIONS and the tests seem to pass?

File: src/postamble_minimal.js [hunk 2]
```diff
@@ -3,6 +3,7 @@
 
 {{{ exportRuntime() }}}
 
+#if hasExportedFunction('_main') // Only if user is exporting a C main(), we will generate a run() function that can be used to launch main.
```
Comment (kripken) [inline src/postamble_minimal.js]: I think `HAS_MAIN` is basically this.

File: src/deps_info.json [hunk 3]
```diff
@@ -94,5 +94,11 @@
   "_embind_register_std_string": ["malloc", "free"],
   "_embind_register_std_wstring": ["malloc", "free"],
   "__syscall192": ["emscripten_builtin_memalign"],
-  "pthread_create": ["malloc", "free", "emscripten_main_thread_process_queued_calls"]
+  "pthread_create": ["malloc", "free", "emscripten_main_thread_process_queued_calls"],
+  "$getTypeName": ["free"],
```
Comment (kripken) [inline src/deps_info.json]: is this one needed? I think deps_info is for C symbols only - the linker that reads that file does not see the JS anyhow. So it must map a C symbol it sees to something that symbol will cause a JS library to require later.

File: src/deps_info.json (hunk 1 above)
Comment (juj) [reply src/deps_info.json]: Oh yeah, added more functions to deps info, which fixed it.

File: src/deps_info.json (hunk 3 above)
Comment (juj) [reply src/deps_info.json]: Right - removed.

File: src/postamble_minimal.js (hunk 2 above)
Comment (juj) [reply src/postamble_minimal.js]: That might be subtly different based on what is assert()ed. I think I'd prefer to leave this as-is, and revisit once I start looking at adding dynamic linking support to MINIMAL_RUNTIME.

Comment (kripken) [reply src/postamble_minimal.js]: Not sure which assertions you mean?

Thinking some more, I think hasExportedFunction is probably better than HAS_MAIN - we should check it exists *and* is exported. How about changing HAS_MAIN to do what hasExportedFunction does, and use HAS_MAIN in this PR?

**Comment by sbc100** [comment]
I seems to remember wanting to warning people when they have a main but forgot to export it..  but I could be wrong. 
 
I think my assumption here was that its probably mis-configured build if you have a `main` in your program but dont' export it.. . if you want to turn you program into a library with no main you probably should exclude main.c from the build completely.   I also often seem(ed) to forget that "EXPORT_FUNCTIONS" on the command line needs to include main so this warning helped me.

File: src/postamble_minimal.js (hunk 2 above)
Comment (juj) [reply src/postamble_minimal.js]: I was referring to the assertion
```js
    assert(!Module['_main'], 'compiled without a main, but one is present. if you added it from JS, use Module["onRuntimeInitialized"]');
```

Also, I am unsure how to change the code with respect to `MAIN_MODULE || SIDE_MODULE || STANDALONE_WASM` checks.

Let me just switch to HAS_MAIN in MINIMAL_RUNTIME, it is probably as close to what is needed here, even if it has a subtle difference.

**Pull request merged**